### ISSUE TITLE
iterator_t and sentinel_t should decay their results

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#161]
currently defined as:

```c++
template<class R>
using iterator_t = decltype(begin(declval<R&>()));
```

Should probably be:

```c++
template<class R>
using iterator_t = decay_t<decltype(begin(declval<R&>()))>;
```

Consider that `iterator_t<tagged_pair<tag::begin(I), tag::end(S)>>` is an alias for `I&`. A reference type is not an iterator.

**Comment by ericniebler** [comment]
Or perhaps the `begin` customization point should always return by value, regardless of whether the user's `begin` overload returns by reference or not. It could be surprising to generic code that the expression `begin(r)` is an lvalue.

**Comment by CaseyCarter** [comment]
> Consider that `iterator_t<tagged_pair<tag::begin(I), tag::end(S)>>` is an alias for `I&`.

I fixed this in my implementation after specifying the customization points, and forgot to specify the change: all of the customization points - those currently specified, that is - should be prvalue expressions.

**Comment by ericniebler** [comment]
Jinx

**Comment by CaseyCarter** [comment]
> Jinx

Heh. It does seem the best approach. I need to figure out an easy way to specify it in the "equivalent expressions" style.

**Comment by CaseyCarter** [comment]
This may also be the solution to maintaining the `Range` / `SizedRange` invariant of pairs/tuples (ericniebler/range-v3#239): access via the customization points is safe, direct access to the members is unsafe and may invalidate the invariant.

**Comment by ericniebler** [comment]
For ericniebler/range-v3#239, I'm just changing the interface to make the data members private for `sized_range` (which I'm renaming to `sized_iterator_range`).

**Comment by CaseyCarter** [comment]
Fixed by fe5163837d973f78ffce55cc614d54c3dc803087.
